corehq/apps/es/README.rst

```diff
@@ -0,0 +1,220 @@
+Elasticsearch App
+=================
+
+Adapter Design
+--------------
+
+The HQ Elastic adapter design came about due to the need for reindexing
+Elasticsearch indexes in a way that is transparent to parts of HQ that write to
+Elasticsearch (e.g. pillowtop).  Reindexing is necessary for making changes to
+index mappings, is a prerequisite to upgrading an Elasticsearch cluster, and is
+also needed for changing low-level index configurations (e.g. sharding).
+
+There is an existing procedure draft that documents the steps that were used on
+one occasion to reindex the ``case_search`` index.  This procedure leveraged a
+custom pillow to "backfill" the cloned index (i.e. initially populated using
+Elasticsearch Reindex API). That procedure only works for a subset of HQ
+Elasticsearch indexes, and is too risky to be considered as an ongoing Elastic
+maintenance strategy. There are several key constraints that an HQ reindexing
+procedure should meet which the existing procedure does not:
+
+- simple and robust
+- performed with standard maintenance practices
+- provides the ability to test and verify the integrity of a new index before it
+  is too late to be rejected
+- allows HQ Elasticsearch index state to remain decoupled from the
+  commcare-cloud codebase
+- is not disruptive -- does not prohibit any other kind of standard maintenance
+  that might come up while the operation is underway
+- is "fire and forget" -- does not require active polling of intermediate state
+  in order to progress the overall operation
+- is practical for third party HQ hosters to use
+
+One way to accomplish these constraints is to implement an "index multiplexing"
+feature in HQ, where Elasticsearch write operations are duplicated across two
+indexes. This design facilitates maintaining two up-to-date versions of any
```

Comment:
One of the reasons to have the back-filling run as a separate process was to prevent any potential lag / performance issues with maintaining the current index. Doing the multiplexing certainly makes the process simpler but do we have any options if the additional operations do cause lag?

Reply:
We need measurements to know how multiplexing will impact the cluster. There is an ongoing discussion with QA on benchmarking HQ Elasticsearch performance with the specific goal of being able to measure any performance degradation as a result of multiplexing. See [QA-3843](https://dimagi-dev.atlassian.net/browse/QA-3843)